src/bar.rs

```diff
@@ -0,0 +1,205 @@
+        len: u64,
+    },
+}
+
+impl PciBarRegion {
+    /// Describe an IO space BAR region.
+    pub fn new_io_region(addr: u32, len: u32) -> Result<Self> {
+        addr.checked_add(len)
+            .ok_or(Error::BarAddressInvalid(addr.into(), len.into()))?;
+        validate_dword_alignment(addr as usize)?;
+        Ok(PciBarRegion::Io { addr, len })
+    }
+
+    /// Describe a memory mapped 32bit BAR region.
+    pub fn new_32bit_mem_region(addr: u32, len: u32) -> Result<Self> {
+        addr.checked_add(len)
+            .ok_or(Error::BarAddressInvalid(addr.into(), len.into()))?;
+        validate_dword_alignment(addr as usize)?;
+        Ok(PciBarRegion::Memory32Bit { addr, len })
+    }
+
+    /// Describe a memory mapped 64bit BAR region.
+    pub fn new_64bit_mem_region(addr: u64, len: u64) -> Result<Self> {
+        addr.checked_add(len)
+            .ok_or(Error::BarAddressInvalid(addr.into(), len.into()))?;
+        validate_dword_alignment(addr as usize)?;
+        Ok(PciBarRegion::Memory64Bit { addr, len })
+    }
+}
+
+#[derive(Copy, Clone)]
+/// Configuration of a Base Address Register.
+pub struct PciBarConfig {
+    /// Index of the BAR.
+    pub index: usize,
+    /// Region referenced by BAR.
+    pub region: PciBarRegion,
+    /// Prefetchable status.
+    pub prefetchable: PciBarPrefetchable,
+}
+
+#[derive(Copy, Clone)]
+/// Configuration of a ROM Base Address Register.
+pub struct PciRomBarConfig {
+    /// Region referenced by BAR.
+    pub region: PciBarRegion,
+    /// Enable bit.
+    pub enable: bool,
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_io_bar_region() {
+        let r = PciBarRegion::new_io_region(0x1000, 4096).unwrap();
+        assert!(
+            r == PciBarRegion::Io {
+                addr: 0x1000,
+                len: 4096
+            }
+        );
+
+        let err_len = u32::MAX - 512;
+        match PciBarRegion::new_io_region(0x1000, err_len) {
+            Err(Error::BarAddressInvalid(0x1000, len)) => {
+                assert_eq!(len, err_len as u64);
+            }
+            _ => assert!(false),
+        }
+
+        match PciBarRegion::new_io_region(0x1003, 4096) {
+            Err(Error::Access(ConfigSpaceAccessError::UnalignedAccess)) => {}
+            _ => assert!(false),
+        }
+    }
+
+    #[test]
+    fn test_32bit_mem_bar_region() {
+        let r = PciBarRegion::new_32bit_mem_region(0x1000, 4096).unwrap();
+        assert!(
+            r == PciBarRegion::Memory32Bit {
+                addr: 0x1000,
+                len: 4096
+            }
+        );
+
+        let err_len = u32::MAX - 512;
+        match PciBarRegion::new_32bit_mem_region(0x1000, err_len) {
+            Err(Error::BarAddressInvalid(0x1000, len)) => {
+                assert_eq!(len, err_len as u64);
+            }
+            _ => assert!(false),
+        }
+
+        match PciBarRegion::new_32bit_mem_region(0x1003, 4096) {
```

Comment:
Can you implement `PartialEq` for all the errors instead? We should only use this verbose way of validating errors only when using `Errors` from crates that don't implement `PartialEq`. Since this `Error` is in our control, we should just make sure to follow the [Rust Interoperability Guidelines](https://rust-lang.github.io/api-guidelines/interoperability.html#types-eagerly-implement-common-traits-c-common-traits), and add all the recommended derives.